### ISSUE TITLE
Stuck in Pre-filling StagingArea

### ISSUE THREAD

**Issue by s7ev3n** [tensorpack/tensorpack#912]
Hi there, 
Thanks for tensorpack !
I am training segmentation model on cityscapes. I write dataflow refering to `get_imagenet_dataflow()`
```
def __iter__(self):
        for img_addr, gt_addr in self.lst:
            img = cv2.cvtColor(cv2.imread(img_addr, cv2.IMREAD_COLOR), cv2.COLOR_BGR2RGB)
            gt = cv2.imread(gt_addr, cv2.IMREAD_GRAYSCALE)
            yield [img, gt]
``` 
And test this dataflow using below code, it prints the numpy array and achieves like 30 it/s(8 cores), and it will suddenly stop at somewhere, like 250/5000.
```
ds = PrefetchDataZMQ(ds, parallel)
    ds = BatchData(ds, batch_size, remainder=False)  
    ds.reset_state()
    print(next(ds.get_data()))
    TestDataSpeed(ds).start()
```
Then run training with `SyncMultiGPUTrainerParameterServer`, the problem is it stuck at Pre-filling StagingArea, showed in below. At the start, CPU is running at 104% with little GPU memory usage, after about 10-15 mins, CPU usage drops and GPU increase, but no computation on GPU with GPU-Util 0%.
I have no idea where I did wrong. Could you give me some insights on this ?? Thanks so much.
```
[0926 11:25:49 @base.py:211] Initializing the session ...
[0926 11:25:49 @base.py:218] Graph Finalized.
[0926 11:25:50 @concurrency.py:37] Starting EnqueueThread QueueInput/input_queue ...
[0926 11:26:01 @param.py:148] [HyperParamSetter] At global_step=0, learning_rate will change to 0.00025000
[0926 11:26:03 @base.py:250] Start Epoch 1 ...
  0%|                                                                                                              |0/371[00:00<?,?it/s]
[0926 11:26:03 @input_source.py:550] Pre-filling StagingArea ...
[0926 11:26:05 @input_source.py:554] 1 element was put into StagingArea on each tower.
```

My environment:
  + Python version: Python 2.7
  + TF version:  tf 1.6.0
  + Tensorpack version: 0.8.9.
  + OS: Ubuntu 16.04
  + Hardware information: E5 2630, 4 1080Ti GPUs.

**Comment by ppwwyyxx** [comment]
Without code I can't tell much about the issue. A stuck can be caused by issues with your data, your model, or some unexpected use of the library (e.g. wrong use of StagingInput). I would recommend to replace your dataflow by `FakeData` to see if your issues still exist or start from a working example to see where things went wrong.

For your dataflow testing: it should run until the end of your dataflow (which is the length of your list divide by the batch size). You can make a dataflow infinite by `RepeatedData(ds, -1)`, which is automatically done if you use a dataflow for training.

**Comment by s7ev3n** [comment]
@ppwwyyxx Thank you so much for the quick reply !! I will check my code, and come back bringing more details.
A little question: What does `print(next(ds.get_data()))` prints out numpy array mean ? To me, it seems the dataflow is running correctly.

**Comment by ppwwyyxx** [comment]
If `FakeData` has the same issue, next step is to replace your model with a trivial model (e.g. `return tf.reduce_sum(input + tf.get_variable('var', shape=[], dtype=tf.float32))`.
If the problem still exists, please post significant parts in the rest of code because I assume there isn't much code left after removing data and model.

**Comment by s7ev3n** [comment]
1.RealData, RealLogits --> Stuck
2.`FakeData`, RealLogits --> Stuck
3.RealData, FakeLogits --> Work !
The model logits part should have bugs ! I will check it ! 
Thanks so much !

**Comment by s7ev3n** [comment]
Hi, I debug my model and find that the dataflow stuck in `BatchNorm()`, in which I use `sync_statistics='nccl'` with `SyncMultiGPUTrainerParameterServer`. When I set `sync_statistics=None`, the training works. 
SyncBN is the most attracting feature for me, what can I do to make it work with SyncBN? Thanks !!

**Comment by ppwwyyxx** [comment]
Could you try SyncMultiGPUTrainerReplicated?

**Comment by s7ev3n** [comment]
In `train.SyncMultiGPUTrainerReplicated(gpus, average=True, mode=None, use_nccl=None)` , should I set `mode='nccl'` or `use_nccl=True` ? I am a little confused.

**Comment by ppwwyyxx** [comment]
It doesn't matter.
With `ParameterServer` I'm still able to use `sync_statistics`.

Could you try a naive model with only one batchnorm layer and please make sure you're using a constant layer name (`BatchNorm('layername', inputs)`) ?

**Comment by s7ev3n** [comment]
I tested `SyncMultiGPUTrainerReplicated` with a naive model and `sync_statistics`, it works ! And my model also works ! 
Thanks so much, I may close this issue now !

**Comment by s7ev3n** [comment]
Hi,
I tried `SyncMultiGPUTrainerParameterServer` with a naive model and constant layername like you said, the code is showed below. It just stuck (I waited for like 5 minutes) and when changed to `SyncMultiGPUTrainerReplicated`, the training process is smoothly running. This seems mysterious.
```
    def get_fake_logits(self, images):
        '''
        Fake logits for testing.
        '''  
        net = Conv2D('fake_conv', images, filters=19, kernel_size=1, strides=1)
        net = BatchNorm('fake_bn', net,  sync_statistics="nccl")
        logits = PReLU('fake_prelu', net)
               
        return logits
```
While all variables in my model is under variable scope 'xxnetTensorpack', training now could run without any warning about syncBN using `SyncMultiGPUTrainerReplicated`, should I delete this variable scope ? 
Or should I try deleting variable scope and `ParameterServer`?

**Comment by ppwwyyxx** [comment]
```diff
diff --git i/examples/basics/cifar-convnet.py w/examples/basics/cifar-convnet.py
index 81e59e2c..2339c098 100755
--- i/examples/basics/cifar-convnet.py
+++ w/examples/basics/cifar-convnet.py
@@ -45,7 +45,8 @@ class Model(ModelDesc):
 
         image = image / 4.0     # just to make range smaller
         with argscope(Conv2D, activation=BNReLU, use_bias=False, kernel_size=3), \
-                argscope([Conv2D, MaxPooling, BatchNorm], data_format=data_format):
+                argscope([Conv2D, MaxPooling, BatchNorm], data_format=data_format), \
+                argscope([BatchNorm], sync_statistics='nccl'):
             logits = LinearWrap(image) \
                 .Conv2D('conv1.1', filters=64) \
                 .Conv2D('conv1.2', filters=64) \
```
The above change on the cifar example (which uses `SyncMultiGPUTrainerParameterServer`) can run well. Does it run on your machine?

**Comment by s7ev3n** [comment]
@ppwwyyxx Sorry for the delay, the cifar example code could run successfully run on my machine. Does it mean that my tensorpack code still has some issues ?
Since I use `SyncMultiGPUTrainerReplicated` trained my model for a few days, the result is much worse than the model I wrote using slim. I first thought the hyper parameter needs modification, but I have tried a few, still no improvement.

**Comment by s7ev3n** [comment]
Thanks! I will check deeper.

**Comment by s7ev3n** [comment]
Hi, @ppwwyyxx I manage to run sync bn with `SyncMultiGPUTrainerParameterServer`.
The issue is probably caused by computing loss part. I used slim's weight decay loss, an auxiliary loss and obtain all losses using `tf.losses.get_total_loss()`, while using tensorpack I added regularize loss to loss collection and  `tf.losses.get_total_loss()`. I now use `tf.add_n()` to add all the losses. I have no idea why this makes sync bn hang.
BTW, I have two little question about the two SyncMulti Trainers: 
1.Does `SyncMultiGPUTrainerParameterServer` synchronize variables per step and `SyncMultiGPUTrainerReplicated` per epoch ? 
2.Does the two trainers sum losses from each tower or average losses across each tower ?
I am new to this cross gpu training. Thanks very much for your answer.

**Comment by ppwwyyxx** [comment]
Adding things to such collections is unsafe, as mentioned in the tutorial: https://tensorpack.readthedocs.io/tutorial/trainer.html

> It may get called multiple times for data-parallel training or inference. As a result:
You’ll need to be careful when modifying global states, e.g. adding ops to collections, setting attributes of a model instance.

So your loss in the last GPU will contain the loss tensor or previous GPUs, that may be the issue.

Your questions:
1. Per-step. They are mathematically equivalent.
2. The docs of these two class has explained it.

**Comment by s7ev3n** [comment]
I did not notice that until now...

I still do not quite understand what happens to losses after reading the docs.
`Data-parallel training in ‘ParameterServer’ mode. It builds one tower on each GPU with shared variable scope. It synchronizes the gradients computed from each tower, averages them and applies to the shared variables.` 
If loss from each tower have shared variable scope, does it mean that it will collect and sum losses from each tower, and then **average** them ? 
Thanks.

**Comment by ppwwyyxx** [comment]
Nothing special happens to losses (if you don't use the collections). It is the gradients that are averaged, not the loss, as said in the above docs you quote.

When you use the collections, the losses in one GPU may contain losses from other GPUs, depend on what you've put in the collection.

**Comment by s7ev3n** [comment]
OK, I get it now. Thanks !
I was considering to modify learning rate compared to on single GPU.

**Comment by s7ev3n** [comment]
Hi, @ppwwyyxx 
May I ask how I save the best validation result model during training?
I tried `infs = [ScalarStats(names='mean_iou', prefix='val')]` and `InferenceRunner(dataset_val, infs)` in callbacks, and also `MaxSaver(monitor_stat='val_mean_iou')`.  While the summary of `val_mean_iou` does not seem right, its value much larger than `mean_iou` and has periodic oscillation.

I guess my method is not right. Thanks !

**Comment by ppwwyyxx** [comment]
The method is correct. As for why the value is periodic, maybe your validation data is not written to produce the same data each time. Or it may be just how your training works: such thing does happen in certain types of training (e.g. when you don't shuffle the training data).

**Comment by s7ev3n** [comment]
Hi, @ppwwyyxx 
I would like to do inference with multiple scales inputs and add those model outputs, i.e. logits, to get final prediction. I tried to build multiple models using `model.get_logits(image)` under `TowerContext`, and feed the scaled input image which is a tf queue tensor, then restore the model variables using `SaverRestore()`, and use `sess.run(final_pred)` to get the final prediction. However, the prediction image is messy like random variables restore. I check the variable names in the saved model, it is the same as building the model, variables are auto-reuse.
The saved model has no problem, as I have tried to do inference using `OfflinePredictor` which is in the tutorial, and it is ok.
Many thanks!

**Comment by ppwwyyxx** [comment]
>  I tried to build multiple models using model.get_logits(image) under TowerContext, and feed the scaled input image which is a tf queue tensor, then restore the model variables using SaverRestore(), and use sess.run(final_pred) to get the final prediction. 

I'm not completely sure what you did and not even sure if any steps above (except for building the model) involves tensorpack. You need to set variable scope to reuse for multiple calls to `get_logits`, and if possible use placeholders to find bugs easier. Print `tf.global_variables` to see what variables you've created, etc.

To report any unexpected issues please use the issue template.

**Comment by MiZhangWhuer** [comment]
> In `train.SyncMultiGPUTrainerReplicated(gpus, average=True, mode=None, use_nccl=None)` , should I set `mode='nccl'` or `use_nccl=True` ? I am a little confused.

@ppwwyyxx @s7ev3n So what's the final options for SyncMultiGPUTrainerReplicated ? I have tried to set it as:
tensorpack.utils.gpu import get_nr_gpu
trainer = SyncMultiGPUTrainerReplicated(max(get_nr_gpu(), 1), mode="nccl") or
trainer = SyncMultiGPUTrainerReplicated(get_nr_gpu(), mode="nccl")

and the BN is set as @ppwwyyxx suggested, but it always failed to conduct synBN. Is that the issues from NVIDIA nccl2 library?
What's the best way to use nccl?  Should I use horovod instead?

**Comment by ppwwyyxx** [comment]
For anyone to better understand and diagnose your issue, please post relevant details following the issue template (click "New Issue" -> "Unexpected Problems / Bugs", or visit [this link to post issues about unexpected problems](https://github.com/tensorpack/tensorpack/issues/new?template=unexpected-problems---bugs.md))
